anonlink/blocking.py

```diff
@@ -0,0 +1,236 @@
+            self._funcs = tuple(args[0])
+        else:
+            self._funcs = args
+
+    @abstractmethod
+    def __call__(self,
+                 dataset_index: int,
+                 record_index: int,
+                 hash_: Sequence[bool]
+                 ) -> Iterable[Hashable]:
+        pass
+
+
+class block_and(_AssociativeBinaryOp):
+    """ Conjunction of multiple blocking functions.
+
+        Records share a block if they share a block in all of the
+        functions.
+
+        :param funcs: Functions whose conjunction we return.
+
+        :return: The blocking function.
+    """
+    __slots__ = ()
+
+    def __call__(self,
+                 dataset_index: int,
+                 record_index: int,
+                 hash_: Sequence[bool]
+                 ) -> Iterable[Hashable]:
+        funcs_eval = _evalf(self._funcs, dataset_index, record_index, hash_)
+        return product(*funcs_eval)
+
+
+class block_or(_AssociativeBinaryOp):
+    """ Disjunction of multiple blocking functions.
+
+        Records share a block if they share a block in any of the
+        functions.
+
+        :param funcs: Functions whose disjunction we return.
+
+        :return: The blocking function.
+    """
+    __slots__ = ()
+
+    def __call__(self,
+                 dataset_index: int,
+                 record_index: int,
+                 hash_: Sequence[bool]
+                 ) -> Iterable[Hashable]:
+        funcs_eval = _evalf(self._funcs, dataset_index, record_index, hash_)
+        funcs_enum = (zip(repeat(i), f) for i, f in enumerate(funcs_eval))
+        return chain.from_iterable(funcs_enum)
+
+
+class bit_blocking:
+    """ Blocking on bits of the hash.
+
+        This blocking type has the advantage that it doesn't require any
+        auxilliary information. Despite this, it is still probabilistic
+        and may decrease recall.
+
+        :param int r: The number of bits to use per block.
+        :param int g: The number of blocks per record.
+        :param seed: Optional seed for pseudorandom number generation.
+            Use to keep the results consistent.
+
+        :return: The blocking function.
+    """
+    
+    __slots__ = '_initial_params', '_hash_len', '_hash_indices'
+
+    def __init__(self,
+                 g: int,
+                 r: int,
+                 *,
+                 seed: Any = None
+                 ) -> None:
+        if g < 1:
+            msg = 'g is expected to be positive but is {}'.format(g)
+            raise ValueError(msg)
+        if r < 1:
+            msg = 'r is expected to be positive but is {}'.format(r)
+            raise ValueError(msg)
+        
+        self._initial_params = g, r, seed  # type: Optional[Tuple[int, int, Any]]
+        self._hash_len = None  # type: Optional[int]
+        self._hash_indices = None  # type: Optional[Sequence[Sequence[int]]]
+
+    def _first_call(self,
+                    hash_len: int
+                    ) -> None:
```

Comment:
I don't understand the point of this function. Why not just calculate `_hash_indices` in the constructor? That way you don't need the `_initial_params` slot at all, and you can delete this function.

Oh, I see, it's because you set `_hash_len` based on the length of the first hash you receive.  I don't much like that style. The hash length will be a system parameter; better (and much simpler) to pass it as a parameter to the constructor.

Reply:
I agree that the implementation would be simpler if we took the hash length as a parameter. However, I’d like to avoid the user of the API having to specify the hash length. This would lead to more complexity in their code.

It makes the API easier to use when it just infers whatever it needs to.

Reply:
Your call.

Personally I think the additional convenience to the caller is pretty minor in this case, while the inconvenience for the maintainer is, let's say, non-trivial.

I'm also not convinced that hiding the dependency on the caller's hash length is an advantage for the caller. For example, I think the caller is more likely to be surprised by the ValueError on L167 if they haven't had to acknowledge the dependence on the hash length.